File: src/fountain.py
import math

from protocol import _imul, splitmix32

_LN2 = 0.6931471805599453

SOLITON_C = 0.1
SOLITON_DELTA = 0.5


def _dlog(x: float) -> float:
    e = 0
    m = x
    while m >= 1.5:
        m /= 2
        e += 1
    while m < 0.75:
        m *= 2
        e -= 1
    z = (m - 1) / (m + 1)
    z2 = z * z
    term = z
    total = 0.0
    for n in range(1, 22, 2):
        total += term / n
        term *= z2
    return e * _LN2 + 2 * total


def _soliton_cdf(k: int) -> list[float]:
    if k == 1:
        return [1.0]
    R = max(1.0, SOLITON_C * _dlog(k / SOLITON_DELTA) * math.sqrt(k))
    spike = min(k, math.ceil(k / R))
    total = 0.0
    cdf = []
    for d in range(1, k + 1):
        rho = 1 / k if d == 1 else 1 / (d * (d - 1))
        tau = 0.0
        if d < spike:
            tau = R / (d * k)
        elif d == spike:
            tau = (R * max(0.0, _dlog(R / SOLITON_DELTA))) / k
        total += rho + tau
        cdf.append(total)
    for i in range(k):
        cdf[i] /= total
    cdf[k - 1] = 1.0
    return cdf


def _to_signed32(v: int) -> int:
    v = v & 0xFFFFFFFF
    return v - 0x100000000 if v >= 0x80000000 else v


def _frame_seed(session_id: int, seq: int) -> int:
    a = (session_id + 1) & 0xFFFFFFFF
    h = _imul(a, 0x9E3779B1) & 0xFFFFFFFF
    b = (seq + 0x85EBCA6B) & 0xFFFFFFFF
    h = (h ^ b) & 0xFFFFFFFF
    h = _to_signed32(h)
    h_unsigned = h & 0xFFFFFFFF
    shifted = h_unsigned >> 13
    h = _to_signed32(h_unsigned ^ shifted)
    h = _imul(h & 0xFFFFFFFF, 0xC2B2AE35) & 0xFFFFFFFF
    h = _to_signed32(h)
    h_unsigned = h & 0xFFFFFFFF
    shifted = h_unsigned >> 16
    h = _to_signed32(h_unsigned ^ shifted)
    return h & 0xFFFFFFFF


def _frame_indices(k: int, cdf: list[float], session_id: int, seq: int) -> list[int]:
    seed = _frame_seed(session_id, seq)
    rnd = splitmix32(seed)
    u = rnd() * 2**-32
    lo, hi = 0, k - 1
    while lo < hi:
        mid = (lo + hi) >> 1
        if cdf[mid] >= u:
            hi = mid
        else:
            lo = mid + 1
    d = min(k, lo + 1)
    if d > k >> 3:
        scratch = list(range(k))
        out = []
        for i in range(d):
            j = i + (rnd() % (k - i))
            scratch[i], scratch[j] = scratch[j], scratch[i]
            out.append(scratch[i])
        return out
    indices: set[int] = set()
    while len(indices) < d:
        indices.add(rnd() % k)
    return list(indices)


def _xor_bytes(a: bytearray, b: bytearray) -> None:
    for i in range(len(a)):
        a[i] ^= b[i]


class LTEncoder:
    def __init__(self, payload: bytes, block_len: int, session_id: int):
        self.block_len = block_len
        self.session_id = session_id
        self.k = max(1, math.ceil(len(payload) / block_len))
        self._blocks: list[bytearray] = []
        padded_len = math.ceil(block_len / 4) * 4
        for b in range(self.k):
            start = b * block_len
            end = min(start + block_len, len(payload))
            block = bytearray(padded_len)
            block[: end - start] = payload[start:end]
            self._blocks.append(block)
        self._cdf = _soliton_cdf(self.k)

    def encode(self, seq: int) -> bytes:
        idx = _frame_indices(self.k, self._cdf, self.session_id, seq)
        padded_len = len(self._blocks[0])
        out = bytearray(padded_len)
        for b in idx:
            _xor_bytes(out, self._blocks[b])
        return bytes(out[: self.block_len])


class _PendingFrame:
    __slots__ = ("idx", "words")

    def __init__(self, idx: set[int], words: bytearray):
        self.idx = idx
        self.words = words


class LTDecoder:
    def __init__(self, k: int, block_len: int, session_id: int, total_len: int):
        self.k = k
        self.block_len = block_len
        self.session_id = session_id
        self.total_len = total_len
        self._padded = math.ceil(block_len / 4) * 4
        self._cdf = _soliton_cdf(k)
        self._solved: list[bytearray | None] = [None] * k
        self._by_block: dict[int, set[_PendingFrame]] = {}
        self._seen: set[int] = set()
        self.solved_count = 0
        self.frames_new = 0
        self.frames_dup = 0

    @property
    def is_complete(self) -> bool:
        return self.solved_count >= self.k

    def add_frame(self, seq: int, block: bytes) -> None:
        if seq in self._seen:
            self.frames_dup += 1
            return
        self._seen.add(seq)
        self.frames_new += 1
        if self.is_complete:
            return
        idx = set(_frame_indices(self.k, self._cdf, self.session_id, seq))
        words = bytearray(self._padded)
        words[: len(block)] = block[: self.block_len]
        for b in list(idx):
            s = self._solved[b]
            if s is not None:
                _xor_bytes(words, s)
                idx.discard(b)
        if not idx:
            return
        if len(idx) == 1:
            self._resolve(next(iter(idx)), words)
            return
        pf = _PendingFrame(idx, words)
        for b in idx:
            if b not in self._by_block:
                self._by_block[b] = set()
            self._by_block[b].add(pf)

    def _resolve(self, b0: int, w0: bytearray) -> None:
        queue: list[tuple[int, bytearray]] = [(b0, w0)]
        while queue:
            b, w = queue.pop()
            if self._solved[b] is not None:
                continue
            self._solved[b] = w
            self.solved_count += 1
            waiting = self._by_block.pop(b, None)
            if not waiting:
                continue
            for pf in waiting:
                _xor_bytes(pf.words, w)
                pf.idx.discard(b)
                if len(pf.idx) == 1:
                    r = next(iter(pf.idx))
                    by_r = self._by_block.get(r)
                    if by_r:
                        by_r.discard(pf)
                    if self._solved[r] is None:
                        queue.append((r, pf.words))

    def assemble(self) -> bytes | None:
        if not self.is_complete:
            return None
        out = bytearray(self.total_len)
        for b in range(self.k):
            start = b * self.block_len
            length = min(self.block_len, self.total_len - start)
            if length > 0 and self._solved[b] is not None:
                out[start : start + length] = self._solved[b][:length]
        return bytes(out)

File: src/static/js/fountain.js
import { splitmix32 } from "./protocol.js";

const LN2 = 0.6931471805599453;

function dlog(x) {
  let e = 0;
  let m = x;
  while (m >= 1.5) { m /= 2; e++; }
  while (m < 0.75) { m *= 2; e--; }
  const z = (m - 1) / (m + 1);
  const z2 = z * z;
  let term = z;
  let sum = 0;
  for (let n = 1; n <= 21; n += 2) {
    sum += term / n;
    term *= z2;
  }
  return e * LN2 + 2 * sum;
}

const SOLITON_C = 0.1;
const SOLITON_DELTA = 0.5;

function solitonCdf(k) {
  const cdf = new Float64Array(k);
  if (k === 1) { cdf[0] = 1; return cdf; }
  const R = Math.max(1, SOLITON_C * dlog(k / SOLITON_DELTA) * Math.sqrt(k));
  const spike = Math.min(k, Math.ceil(k / R));
  let total = 0;
  for (let d = 1; d <= k; d++) {
    const rho = d === 1 ? 1 / k : 1 / (d * (d - 1));
    let tau = 0;
    if (d < spike) tau = R / (d * k);
    else if (d === spike) tau = (R * Math.max(0, dlog(R / SOLITON_DELTA))) / k;
    total += rho + tau;
    cdf[d - 1] = total;
  }
  for (let i = 0; i < k; i++) cdf[i] /= total;
  cdf[k - 1] = 1;
  return cdf;
}

function frameSeed(sessionId, seq) {
  let h = (Math.imul(sessionId + 1, 0x9e3779b1) ^ (seq + 0x85ebca6b)) | 0;
  h = Math.imul(h ^ (h >>> 13), 0xc2b2ae35);
  return (h ^ (h >>> 16)) | 0;
}

function frameIndices(k, cdf, sessionId, seq) {
  const rnd = splitmix32(frameSeed(sessionId, seq));
  const u = rnd() * 2 ** -32;
  let lo = 0, hi = k - 1;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (cdf[mid] >= u) hi = mid;
    else lo = mid + 1;
  }
  const d = Math.min(k, lo + 1);
  if (d > k >> 3) {
    const scratch = new Uint32Array(k);
    for (let i = 0; i < k; i++) scratch[i] = i;
    const out = new Array(d);
    for (let i = 0; i < d; i++) {
      const j = i + (rnd() % (k - i));
      const t = scratch[i];
      scratch[i] = scratch[j];
      scratch[j] = t;
      out[i] = scratch[i];
    }
    return out;
  }
  const set = new Set();
  while (set.size < d) set.add(rnd() % k);
  return [...set];
}

function xorInto(dst, src) {
  for (let i = 0; i < dst.length; i++) dst[i] = (dst[i] ^ src[i]) >>> 0;
}

export class LTEncoder {
  constructor(payload, blockLen, sessionId) {
    this.blockLen = blockLen;
    this.sessionId = sessionId;
    this.k = Math.max(1, Math.ceil(payload.length / blockLen));
    this._words = Math.ceil(blockLen / 4);
    this._blocks = new Uint32Array(this.k * this._words);
    const bytes = new Uint8Array(this._blocks.buffer);
    for (let b = 0; b < this.k; b++) {
      const src = payload.subarray(b * blockLen, Math.min((b + 1) * blockLen, payload.length));
      bytes.set(src, b * this._words * 4);
    }
    this._cdf = solitonCdf(this.k);
  }

  encode(seq) {
    const idx = frameIndices(this.k, this._cdf, this.sessionId, seq);
    const out = new Uint32Array(this._words);
    for (const b of idx) {
      const off = b * this._words;
      for (let w = 0; w < this._words; w++) out[w] = (out[w] ^ this._blocks[off + w]) >>> 0;
    }
    return new Uint8Array(out.buffer, 0, this.blockLen);
  }
}

export class LTDecoder {
  constructor(k, blockLen, sessionId, totalLen) {
    this.k = k;
    this.blockLen = blockLen;
    this.sessionId = sessionId;
    this.totalLen = totalLen;
    this._words = Math.ceil(blockLen / 4);
    this._cdf = solitonCdf(k);
    this._solved = new Array(k).fill(null);
    this._byBlock = new Map();
    this._seen = new Set();
    this.solvedCount = 0;
    this.framesNew = 0;
    this.framesDup = 0;
  }

  get isComplete() {
    return this.solvedCount >= this.k;
  }

  addFrame(seq, block) {
    if (this._seen.has(seq)) { this.framesDup++; return; }
    this._seen.add(seq);
    this.framesNew++;
    if (this.isComplete) return;

    const idx = new Set(frameIndices(this.k, this._cdf, this.sessionId, seq));
    const words = new Uint32Array(this._words);
    new Uint8Array(words.buffer).set(block.subarray(0, this.blockLen));
    for (const b of [...idx]) {
      const s = this._solved[b];
      if (s) { xorInto(words, s); idx.delete(b); }
    }
    if (idx.size === 0) return;
    if (idx.size === 1) {
      this._resolve(idx.values().next().value, words);
      return;
    }
    const pf = { idx, words };
    for (const b of idx) {
      let set = this._byBlock.get(b);
      if (!set) { set = new Set(); this._byBlock.set(b, set); }
      set.add(pf);
    }
  }

  _resolve(b0, w0) {
    const queue = [[b0, w0]];
    while (queue.length > 0) {
      const [b, w] = queue.pop();
      if (this._solved[b]) continue;
      this._solved[b] = w;
      this.solvedCount++;
      const waiting = this._byBlock.get(b);
      if (!waiting) continue;
      this._byBlock.delete(b);
      for (const pf of waiting) {
        xorInto(pf.words, w);
        pf.idx.delete(b);
        if (pf.idx.size === 1) {
          const r = pf.idx.values().next().value;
          this._byBlock.get(r)?.delete(pf);
          if (!this._solved[r]) queue.push([r, pf.words]);
        }
      }
    }
  }

  assemble() {
    if (!this.isComplete) return null;
    const out = new Uint8Array(this.totalLen);
    for (let b = 0; b < this.k; b++) {
      const start = b * this.blockLen;
      const len = Math.min(this.blockLen, this.totalLen - start);
      if (len > 0) out.set(new Uint8Array(this._solved[b].buffer, 0, len), start);
    }
    return out;
  }
}
